src/webnn/native/dml/UtilsDML.h

```diff
@@ -0,0 +1,135 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef WEBNN_NATIVE_UTILSDML_H_
+#define WEBNN_NATIVE_UTILSDML_H_
+
+#define DML_TARGET_VERSION_USE_LATEST 1
+
+#include <dxgi1_6.h>
+#include <webnn/webnn_cpp.h>
+#include <wrl\client.h>
+#include <deque>
+#include <map>
+#include <unordered_map>
+#include <vector>
+
+#include "DirectML.h"
+#include "common/Assert.h"
+#include "common/Log.h"
+#include "webnn/native/NamedOutputs.h"
+#include "webnn/native/webnn_platform.h"
+
+namespace webnn::native::dml {
+#define ASSERT_IF_FAILED(hr)                                                                \
+    if (FAILED(hr) && hr != S_FALSE) {                                                      \
+        dawn::ErrorLog() << "Failed to do " << #hr << " Return HRESULT " << std::hex << hr; \
+        DAWN_ASSERT(0);                                                                     \
+    }
+
+#define RETURN_IF_FAILED(hr)                                                                \
+    if (FAILED(hr) && hr != S_FALSE) {                                                      \
+        dawn::ErrorLog() << "Failed to do " << #hr << " Return HRESULT " << std::hex << hr; \
+    }
+
+    using namespace Microsoft::WRL;
+
+    // Represent the DirectML tensor description.
+    struct TensorDesc {
+        std::vector<UINT> dimensions = {};
+        std::vector<UINT> strides = {};
+        DML_BUFFER_TENSOR_DESC bufferDesc = {};
+    };
+
+    enum NodeType { NonConstantInput, ConstantInput, Intermediate };
+
+    // Represent the information of the graph's nodes.
+    struct NodeBase {
+        virtual ~NodeBase() = default;
+        DML_TENSOR_DESC outputTensorDesc = {};
+        std::string name = "";
+        NodeType type = NodeType::Intermediate;
+    };
+
+    // Only represent the information of the input nodes.
+    struct InputNode final : public NodeBase {
+        ~InputNode() override = default;
+        // Indicate the index of the graph's input.
+        size_t inputIndex = 0;
+        void const* buffer = nullptr;
+        size_t byteLength = 0;
+    };
+
+    // Represent the information of the intermediate and output nodes.
+    struct Node final : public NodeBase {
+        ~Node() override = default;
+        uint32_t nodeIndex = 0;
+        uint32_t outputNodeIndex = 0;
+    };
+
+    // Describe a graph of DirectML operators used to compile a combined, optimized operator.
+    class GraphBuilder {
+      public:
+        GraphBuilder(ComPtr<IDMLDevice> device) : mDevice(device) {
```

Comment:
unit: use `mDevice(std::move(device))` to avoid the unnecessary ref-count.

Reply:
Already change to use IDMLDevice* device as parameter